File: 0221-maximal-square/0221-maximal-square.java
class Solution {
    int dp[][];
    int maximalSquareSize;
    public int maximalSquare(char[][] matrix) {
        
        int r = matrix.length;
        int c = matrix[0].length;
        
        dp = new int[r][c];
        
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                dp[i][j] = -1;
            }
        }
        
        solve(matrix,0,0);
        
        //return (int) Math.pow(maximalSquareSize, 2);
        return maximalSquareSize * maximalSquareSize;
    }
    
    public int solve(char[][] mat, int m, int n){
        
        if(m >= mat.length || n >= mat[0].length) return 0;

        if(dp[m][n] != -1) return dp[m][n];
        
        int right = solve(mat, m, n + 1);
        int diagonal = solve(mat, m + 1, n + 1);
        int down = solve(mat, m + 1, n);
        
        if(mat[m][n] == '1'){
            dp[m][n] = 1 + Math.min(right, Math.min(diagonal, down));
            maximalSquareSize = Math.max(maximalSquareSize, dp[m][n]);
            return dp[m][n];
        }else{
            return dp[m][n] = 0;
        }
         
    }
}

File: 0221-maximal-square/0221-maximal-square.cpp
class Solution {

public:
    int maxi = 0;
    int recursive(int i,int j, vector<vector<char>> &matrix, vector<vector<int>> &dp){
        if(i>=matrix.size() || j>=matrix[0].size()) return 0;
        if(dp[i][j]!=-1) return dp[i][j];
        int right = recursive(i,j+1,matrix,dp);
        int diagonal = recursive(i+1,j+1,matrix,dp);
        int down = recursive(i+1,j,matrix,dp);
        if(matrix[i][j]=='1'){
            int x = 1+min(right,min(diagonal,down));
            maxi = max(maxi,x);
            return dp[i][j]=x;
        } 
        return dp[i][j]=0;
    }
    
    int maximalSquare(vector<vector<char>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        vector<vector<int>> dp(n,vector<int>(m,-1));
        recursive(0,0,matrix,dp);
        return maxi*maxi;
    }
};